### DISCUSSION TITLE
How to chain prompts in teams-ai version 1.0.0 (MS Teams AI Bot app)?

### DISCUSSION THREAD

**Discussion by donatas-xyz** [microsoft/teams-sdk#1322] [Q&A]
I'm raising this question here as suggested by @corinagum. 

In `@microsoft/teams-ai: ^0.1.0-preview.2` I used to use `AI<ApplicationTurnState>.chain(context: TurnContext, state: ApplicationTurnState, prompt?: string | PromptTemplate, options?: Partial<AIOptions<ApplicationTurnState>>): Promise<...> Chains into another prompt and executes the plan that is returned.`.

Example:

```typescript 
await app.ai.chain(context, state, 'generateTenancyAgreement');
```
However in the new `@microsoft/teams-ai: ^1.0.0` I cannot for the life of me find anything like that? The closest I've got is `ActionPlanner<TurnState<DefaultConversationState, DefaultUserState, DefaultTempState>>.completePrompt<string>(context: TurnContext, memory: Memory, prompt: string | PromptTemplate, validator?: PromptResponseValidator<string>): Promise<...> Completes a prompt using an optional validator.`.

Example:

```typescript
const response = await planner.completePrompt(context, state, 'generateTenancyAgreement');
```

I've looked at all [the updated examples](https://github.com/microsoft/teams-ai) and couldn't see anything in regards of this functionality?

I think I've managed to address all the other breaking changes but this one. 

As you can see from the example below, the `generateTenancyAgreement` prompt is chained onto `chat` prompt like so: 
general chat starts > "generate tenancy agreement" > bot requests property address following instructions inside the `chat` prompt > `generateTenancyAgreement.ts` is executed following `chat` prompt logic still > `await app.ai.chain(context, state, 'generateTenancyAgreement');` is used to chain `src\prompt\generateTenancyAgreement` with it's own parameters and prompt instructions > `src\prompt\generateTenancyAgreement\skprompt.txt` are being followed and expected result returned. 

How would you achieve the same in version 1.0.0? 

**EDIT:** It would seem I can use `planner.completePrompt(context, state, 'generateTenancyAgreement')` instead of `chain()`. It may be that the bug #1130 was throwing me off massively in my efforts or something changed slightly since teams-ai 1.3. 

###### `config.json`
```json
{
  "schema": 1,
  "description": "AI Bot",
  "type": "completion",
  "completion": {
    "max_tokens": 2000,
    "temperature": 0.8,
    "top_p": 0.0,
    "presence_penalty": 0.0,
    "frequency_penalty": 0.0,
    "stop_sequences": [
      "Person:",
      "Bot:"
    ]
  }
}
```
###### `skprompt.txt` (simplified)
```text 
#### START PROMPT INSTRUCTIONS

## Prompt grounding
- You are an AI assistant that helps people find information. 

#### END PROMPT INSTRUCTIONS

#### START ACTION INSTRUCTIONS 

Only the following actions and functions are supported:
- {"type":"DO","action":"generateTenancyAgreement","entities":{"operation": "generate", "address": "<propertyAddress>"}}

When to use actions and functions:
- Use `generateTenancyAgreement` to generate an assured shorthold tenancy agreement when the Person asks to create a tenancy agreement 
- Always try to use actions or funtions first, before returning model's response 

What to do after action and functions have been executed: 
- Return the action results to the person. 
- If the chat doesn't match the action then just use standard chat. 
- The Assistant must return the following JSON structure if it matches an action:
{"type":"plan","commands":[{"type":"DO","action":"<name>","entities":{"<name>":<value>}},{"type":"SAY","response":"<response>"}]}

#### END ACTION INSTRUCTIONS 

#### START ACTION EXAMPLES

Person: generate a tenancy agreement 
Bot: What is the property address? 
Person: {{$input}} 
Bot: DO generateTenancyAgreement operation="generate" address="{{$input}}" THEN SAY Can I help you with anything else? 

#### END ACTION EXAMPLES

{{$history}}
Person: {{$input}}
Bot: 
```
###### `generateTenancyAgreement.ts` 
```typescript
export function generateTenancyAgreement(app: Application<ApplicationTurnState>): void 
{
    app.ai.action("generateTenancyAgreement", async (context: TurnContext, state: ApplicationTurnState, data: ITenancyAgreement) => 
    {
        const action = data.operation!.toLowerCase();
        const address = data.address!;
        state.temp.value.propertyAddress = address;
      
        switch (action) 
        {
            case 'generate':
                await context.sendActivity(`&#8987; Generating a Tenancy Agreement for **'${address}'**...`);
                await app.ai.chain(context, state, 'generateTenancyAgreement');
                return false;
            default:
                await context.sendActivity(`Oops, my developer hasn't implemented the ${action} path of 'generateTenancyAgreement' yet!`);
                return true;
        }
    });
}
```

###### `config.json`
```json
{
	"schema": 1,
	"description": "Generate Tenancy Agreement.",
	"type": "completion",
	"completion": {
		"max_tokens": 4000,
		"temperature": 0.4,
		"top_p": 0.0,
		"presence_penalty": 0.0,
		"frequency_penalty": 0.0,
		"stop_sequences": [
			"Person:",
			"Bot:"
		]
	},
	"input": {
		"parameters": [
			{
				"name": "history",
				"description": "The history of the conversation.",
				"defaultValue": ""
			},
			{
				"name": "address",
				"description": "Property address.",
				"defaultValue": ""
			}
		]
	}
}
```
###### `skprompt.txt` (simplified)
```text 
You are a Legal Advisor at ... 
Your name is Bot. 
Your job is to provide guidance on tenancy agreements. 

You have been asked by the Person to generate Assured Shorthold Tenancy Agreement for your client to read and sign.

ASSUMPTIONS:
Use following **information** to generate the agreement: 
- The agreement is for a residential property. 
- The agreement is for a fixed term of 12 months. 
- The property address is {{$propertyAddress}}

STEPS:
1. Use the **information** provided and the **template** below to generate the agreement: 
	1.1.1. Paragraph headers inside the "agreementBody" node should start a numbered list 

{agreementBodyJsonString-Template}

2. Using the information inside the **template** generate a human readable **text document** ready to be printed out 
	2.1. Format the "header" nodes as paragraph headers 
		2.1.1. Paragraph headers should be in bold text 
	3. **Display** the **text document** 

Let's do it in steps 

RULES: 
- Show **only** the information in the **text document** and **nothing** else 

Bot: **text document**
```

- [StackOverflow query](https://stackoverflow.com/questions/77808829/how-to-chain-prompts-in-teams-ai-version-1-0-0-ms-teams-ai-bot-app?noredirect=1#comment137574945_77808829) 
- [GitHub query](https://github.com/microsoft/teams-ai/discussions/1060#discussioncomment-8184835)

**Comment by corinagum** [comment]
Hi @donatas-xyz, thank you for posting! We have indeed deprecated the `chain()` feature, but chaining actions should still be possible [using `app.ai.doAction`](https://github.com/microsoft/teams-ai/blob/main/js/samples/04.ai.d.chainedActions.listBot/src/index.ts#L149)
![image](https://github.com/microsoft/teams-ai/assets/14900841/fc674ab6-f74c-4bb3-8f47-b1314e5c206e)

Could you share your `actions.json` file? In case the file hasn't been created yet, I'm wondering if that or some formatting issue may be interfering with the bot's functionality.

**Reply by donatas-xyz** [reply]
Hi @corinagum, 

I don't have my code handy, but I believe you are talking about actions and I'm talking about prompts. 

Executing an action is not an issue for me, however what I used to be able to do with the `chain()` function is to run different prompt instructions. 

For example, say you have 2 prompts - one general one called "chat" and another called "generateDocument", both sitting in their own prompt folders with their own `config.json` and `skprompt.txt` files like so: 
```text
|_src\prompts
  |_chat
    |_config.json 
    |_skprompt.txt 
  |_generateDocument 
    |_config.json 
    |_skprompt.txt   
``` 
As you can see, both *.json and both *.txt files can have very different parameters and prompts. 

For example, "chat" prompt can say that your bot is "_a friendly AI helper ready to answer any questions_", and therefore its temperature be set to, say, 0.9, whereas "generateDocument" prompt would say your bot is "_a legal advisor specialising in property law and drafting tenancy agreements_", and therefore its temperature would be more like 0.4 to avoid hallucinations. 

When you start your bot, you normally start using general "chat" prompt, but using `chain()` you could easily switch from "chat" to "generateDocument" prompt as soon as you execute relevant referenced action within "chat" prompt (let's call it `app.ai.action('generateTenancyAgreement'...)`), like I've described in my first post - so your friendly AI bot all of the sudden becomes a lawyer or a maths teacher or a senior developer etc :) 

I'm also pretty sure, that the scenario I've described in my first post would not work with Monologue augmentation type. 

I hope this makes more sense. 

Thank you, @corinagum!

**Reply by corinagum** [reply]
Hi @donatas-xyz, my sincere apologies! You are quite correct and I have gotten my wires crossed. Let me circle back and get back to you asap.

**Reply by corinagum** [reply]
@donatas-xyz, Thanks for your patience! To re-iterate, prompt chaining was removed from the SDK before our release to GA. Pre-GA, everything was considered experimental and not guaranteed to remain in the codebase. At the moment, there's no direct replacement like what you're discussing above in your clarification - the way we have now to affect chaining is through actions, as described in my original reply. 

To get more insight on why `chain()` was removed, @Stevenic would be able to provide more information. Steve, if you see this, could you provide details on the decision for removing `chain()`?

Otherwise, if there is enough customer interest in bringing back `chain()`, it's certainly something we can look into adding to our roadmap. As of right now though, there's no plan for a direct replacement. 

If there are other developers who are interested in prompt chaining like `chain()` used to do, let's please use this discussion item as a sounding board for people to voice their thoughts. If there is enough engagement, I can return to my team to discuss the option of bringing it back.

Sorry this isn't the response you are looking for, but I hope that clarifies things!

**Reply by lifan0127** [reply]
Hi I am late in the game so I don't have experience with `app.ai.chain`. I have used the `planner.completePrompt` approach as mentioned by @donatas-xyz to achieve similar functionality, but it requires me to handle the response manually with `doAction` etc. It seems that `app.ai.chain` is more straightforward /elegant.

By the way, @corinagum I believe we are missing an equivalent `do_action` method for the Python implementation.

**Reply by kevmcdonk** [reply]
I was using Chain in the sample for Graph SSO and found it was a great way for using RAG pattern and then returning it with a prompt. Is there really no way of doing this now?

https://github.com/kevmcdonk/TeamsAI-GraphSample/blob/main/index.ts

**Reply by corinagum** [reply]
Hi @donatas-xyz, @lifan0127, @kevmcdonk,

Because there's been a lot of discussion about `chain()` since it got deprecated, I brought it up with my team. Currently there is no plan to bring chain back since `doAction` is available, however, if any of you are still keen on the feature, please feel free to file 1 (one) feature request to bring chain back for prompt kickoff. 

I can't guarantee it will be brought back, but the more of you who +1 the feature request, the more weight is given for consideration when planning our roadmap :) If you can spread the word to get upvotes/reactions that of course helps too.

@lifan0127, I saw your comment about `do_action` for Python and filed a work item for that. Thanks for the heads up.

**Comment by autonomous-agent-7** [comment]
📋 **Team Lead Routing**

This feature request has been received and routed to @project-manager for triage and prioritization.

_Context: Copilot provider adapter port from opencode._
